### ISSUE TITLE
Conflicts between mscorlib 4.0.0.0 and System.Runtime 4.0.20.0

### ISSUE THREAD

**Issue by CuddleBunny** [aspnet/dnx#2243]
For the last few days I haven't been able to build any beta5 or beta6-* projects. I get over 1,500 weird errors like `Predefined type 'System.Void' is not defined or imported` but I think there is a conflict between mscorlib and System.Runtime:

    The type 'AssemblyFileVersionAttribute' exists in both 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' and 'System.Runtime, Version=4.0.20.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'

Here is `dnvm list`:

    Active Version           Runtime Architecture Location                         Alias
    ------ -------           ------- ------------ --------                         -----
           1.0.0-beta4       clr     x64          C:\Users\jdovishaw\.dnx\runtimes
           1.0.0-beta4       clr     x86          C:\Users\jdovishaw\.dnx\runtimes
           1.0.0-beta4       coreclr x64          C:\Users\jdovishaw\.dnx\runtimes
           1.0.0-beta4       coreclr x86          C:\Users\jdovishaw\.dnx\runtimes
           1.0.0-beta5       clr     x86          C:\Users\jdovishaw\.dnx\runtimes
      *    1.0.0-beta6-12207 clr     x86          C:\Users\jdovishaw\.dnx\runtimes default

Building projects with beta4 does seem to work however.

**Comment by anurse** [comment]
Can you post the `project.json` you are trying to build?

**Comment by CuddleBunny** [comment]
Sure, I have a Shared Project:

	{
		"version": "1.0.0-*",
		"description": "",
		"authors": [ "Josh Dovishaw" ],

		"dependencies": {
			"AutoMapper": "3.3.1",
			"EntityFramework": "7.0.0-beta6-*",
			"EntityFramework.Commands": "7.0.0-beta6-*",
			"EntityFramework.Core": "7.0.0-beta6-*",
			"EntityFramework.SqlServer": "7.0.0-beta6-*",
			"Microsoft.AspNet.Hosting.Abstractions": "1.0.0-beta6-*",
			"Microsoft.AspNet.Http": "1.0.0-beta6-*",
			"Microsoft.AspNet.Http.Extensions": "1.0.0-beta6-*",
			"Microsoft.AspNet.Identity.EntityFramework": "3.0.0-beta6-*",
			"Microsoft.AspNet.Mvc": "6.0.0-beta6-*",
			"Microsoft.AspNet.Providers.Core": "2.0.0",
			"Microsoft.AspNet.Server.IIS": "1.0.0-beta6-*",
			"Microsoft.AspNet.Session": "1.0.0-beta6-*",
			"Microsoft.Framework.CodeGenerators.Mvc": "1.0.0-beta6-*",
			"Microsoft.Framework.Configuration.Json": "1.0.0-beta6-*",
			"Microsoft.Framework.OptionsModel": "1.0.0-beta6-*",
			"Microsoft.Net.Http": "2.2.29",
			"RavenDB.Embedded": "3.0.3685-Unstable",
			"System.IO.Compression.ZipFile": "4.0.0-beta-*",
			"System.Threading": "4.0.10-beta-*"
		},

		"commands": {
			"gen": "Microsoft.Framework.CodeGeneration",
			"ef": "EntityFramework.Commands"
		},

		"frameworks": {
			"dnx451": {
				"frameworkAssemblies": { "System.ComponentModel.DataAnnotations": "" }
			}
		}
	}

and my main project:

	{
		"webroot": "wwwroot",
		"version": "1.0.0-*",
		"authors": [ "Josh Dovishaw" ],

		"dependencies": {
			"AutoMapper": "3.3.1",
			"EntityFramework": "7.0.0-beta6-*",
			"EntityFramework.Core": "7.0.0-beta6-*",
			"EntityFramework.SqlServer": "7.0.0-beta6-*",
			"jQuery": "2.1.3",
			"LMS.Shared": "",
			"Microsoft.AspNet.Diagnostics": "1.0.0-beta6-*",
			"Microsoft.AspNet.Diagnostics.Entity": "7.0.0-beta6-*",
			"Microsoft.AspNet.Identity": "3.0.0-beta6-*",
			"Microsoft.AspNet.Identity.EntityFramework": "3.0.0-beta6-*",
			"Microsoft.AspNet.Mvc": "6.0.0-beta6-*",
			"Microsoft.AspNet.Mvc.TagHelpers": "6.0.0-beta6-*",
			"Microsoft.AspNet.Razor": "4.0.0-beta6-*",
			"Microsoft.AspNet.Server.IIS": "1.0.0-beta6-*",
			"Microsoft.AspNet.Server.WebListener": "1.0.0-beta6-*",
			"Microsoft.AspNet.Session": "1.0.0-beta6-*",
			"Microsoft.AspNet.StaticFiles": "1.0.0-beta6-*",
			"Microsoft.Framework.CodeGenerators.Mvc": "1.0.0-beta6-*",
			"Microsoft.Framework.Configuration": "1.0.0-beta6-*",
			"Microsoft.Framework.Configuration.Json": "1.0.0-beta6-*",
			"Microsoft.Framework.DependencyInjection": "1.0.0-beta6-*",
			"Microsoft.Framework.DependencyInjection.Abstractions": "1.0.0-beta6-*",
			"Microsoft.Net.Http": "2.2.29",
			"RavenDB.Embedded": "3.0.3685-Unstable",
			"System.IO.Compression.ZipFile": "4.0.0-beta-*",
			"System.Threading": "4.0.10-beta-*"
		},

		"commands": {
			"web": "Microsoft.AspNet.Hosting --server Microsoft.AspNet.Server.WebListener --server.urls http://localhost:5000",
			"gen": "Microsoft.Framework.CodeGeneration",
			"ef": "EntityFramework.Commands"
		},

		"frameworks": {
			"dnx451": {

			}
		},

		"publishExclude": [
			"node_modules",
			"bower_components",
			"**.xproj",
			"**.user",
			"**.vspscc"
		],
		"exclude": [
			"wwwroot",
			"node_modules",
			"bower_components"
		]
	}

**Comment by anurse** [comment]
Can you try this temporary workaround: Add a direct dependency to your dependencies list on `System.Runtime` version `4.0.10-beta-*` (note that's **10** NOT 20). This shouldn't be required, but if it fixes the issue then I have a pretty good idea where to look next :).

**Comment by CuddleBunny** [comment]
Now I only get 100 errors instead of 1,500 but I have a feeling that is just because it stopped earlier in the build process. Newer seemingly related errors are like:

    The dependency System.Collections 4.0.10-beta-22231 in project LMS.Shared does not support framework DNX,Version=v4.5.1.

    The predefined type 'Task' is defined in multiple assemblies in the global alias; using definition from 'mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'

I am now also getting an error from Nuget:

    Could not find a part of the path 'C:\Users\jdovishaw\.dnx\packages\System.Threading\4.0.0\System.Threading.4.0.0.nupkg.sha512'.

**Comment by anurse** [comment]
Great info, thanks! I think I have an idea what the underlying problem is, I'll try some things on my end and see what I find out.

**Comment by muratg** [comment]
@anurse we can move it out if the fix turns out to be not straightforward

**Comment by anurse** [comment]
Ok, the issue is `System.IO.Compression.ZipFile`. It is not compatible with `dnx451`. If it worked previously, this was a bug at that time. I would expect it depends upon new functionality in the .NET 4.6 framework, but someone from the CLR team (cc'd below) can confirm that.

You could try switching to the latest `beta6` builds and change your app to target `dnx46` which would mean you are targeting .NET 4.6.

/cc @ericstj and @davidfowl  to confirm my interpretation is correct. `System.IO.Compression.ZipFile` depends on `System.Runtime 4.0.20` which is not available on .NET < 4.6

There isn't really a DNX issue here any more, so I'm going to close this, but please do re-open it if you still have issues and keep commenting on the thread if you need more information!

**Comment by anurse** [comment]
I confirmed that upgrading to `beta6-12207` or higher (which is on the `aspnetvnext` feed) and changing to `dnx46` no longer reproduces the error in my test. I don't have your full code so if that doesn't work for you, let me know what you are seeing.

**Comment by ericstj** [comment]
Net46 is where all the new contracts light up, on net451 you can use a framework assembly ref to System.IO.Compression.Filesystem.

**Comment by CuddleBunny** [comment]
Interesting. I've been using ZipFile in 451 since like April, it only just broke this week I think. I tried changing targets to 452 and 46 earlier this week because I saw them in some of the commit notes and it seems to build error free. I could not however, figure out how to get a dnx46 compatible version of the runtime, running the app gives me the following runtime error:

	Server Error in '/' Application.
	The current runtime target framework is not compatible with 'LMS.Web'.

	Current runtime Target Framework: 'DNX,Version=v4.5 (dnx45)'
	 Type: CLR
	 Architecture: x86
	 Version: 1.0.0-beta6-12207

	Please make sure the runtime matches a framework specified in project.json
	Description: An unhandled exception occurred during the execution of the current web request. Please review the stack trace for more information about the error and where it originated in the code.

	Exception Details: System.InvalidOperationException: The current runtime target framework is not compatible with 'LMS.Web'.

	Current runtime Target Framework: 'DNX,Version=v4.5 (dnx45)'
	 Type: CLR
	 Architecture: x86
	 Version: 1.0.0-beta6-12207

	Please make sure the runtime matches a framework specified in project.json

Both `project.json` files have 

    "frameworks": {
		"dnx46": { ...

Is there something else I have to do to target the newer framework?

**Comment by anurse** [comment]
There was an issue in `Microsoft.AspNet.Server.IIS` that was exposed by recent DNX changes. That's been fixed in `1.0.0-beta6-12044` of `Microsoft.AspNet.Server.IIS`. Versions later than that have already been published to `aspnetvnext` so can you try re-restoring and see if you get a new enough version? You can find the exact version you got by running `dnu list` after the restore.

**Comment by anurse** [comment]
As @ericstj said you can also go back to `dnx451` by changing the `System.IO.Compression.ZipFile` reference to a framework assembly (you still need to get the updated IIS Server package as I mentioned above)

**Comment by CuddleBunny** [comment]
Alright, I would definitely prefer to be on 4.6. My project currently has `Microsoft.AspNet.Server.IIS (1.0.0-beta6-12057)` so I don't think that is the problem. I don't see any target framework settings in the project properties or if I open the xproj in a text editor. 

When I went back to 4.5.1, I also had to remove `System.Xml.XmlDocument` in addition to `System.IO.Compression.ZipFile` to get it to build.
